src/Database/Esqueleto/PostgreSQL/Pgcrypto.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | This module contains functions specific to the [pgcrypto](https://www.postgresql.org/docs/current/pgcrypto.html) module
+-- 
+-- @since: 3.5.1
+
+module Database.Esqueleto.PostgreSQL.Pgcrypto 
+    (HashAlgorithm(..),
+     toCrypt,
+     fromCrypt,
+    ) where
+
+import Database.Esqueleto.Internal.Internal 
+import Database.Esqueleto.Experimental (toPersistValue)
+
+{- pgcrypto hashing algorithms
+see: https://www.postgresql.org/docs/current/pgcrypto.html
+
+/Requires/ the pgcrypto module.
+
+@since 3.5.1
+-}
+data HashAlgorithm = BF | MD5 | XDES | DES
+  deriving (Eq, Show)
+
+{- | (@crypt()@) Calculate a crypt-like hash from the provided password
+
+/Requires/ the pgcrypto module.
+
+/WARNING/: Using `toCrypt` may leak sensitive data via logging. Filtering logs in production environments
+when using `toCrypt`, such as using `filterLogger` on `monad-logger` based stacks is highly advised.
+
+example:
+@
+share
+    [mkPersist sqlSettings]
+    [persistLowerCase|
+    UserAccount json
+        name T.Text
+        UniqueName name
+        passwordHash T.Text
+        deriving Show Read Eq
+
+insertSelect $ do
+    pure $
+        UserAccount
+            <# val "username"
+            <&> toCrypt BF "1234password"
+@
+
+-- @since 3.5.1
+-}
+toCrypt :: SqlString s => HashAlgorithm -> s -> SqlExpr (Value s)
+toCrypt algorithm pass =
+    let alg = case algorithm of
+            BF -> "bf"
+            MD5 -> "md5"
+            XDES -> "xdes"
+            DES -> "des"
+    in ERaw noMeta $ \_ _ -> ("crypt (?, gen_salt('" <> alg <> "'))", [toPersistValue pass])
+
+{- | (@crypt()@) Retrieve a hashed password
+
+/Requires/ the pgcrypto module.
+
+example:
+@
+share
+    [mkPersist sqlSettings]
+    [persistLowerCase|
+    UserAccount json
+        name T.Text
+        UniqueName name
+        passwordHash T.Text
+        deriving Show Read Eq
+
+
+login name pwd = select $ do
+    user <- from $ Table UserAccount
+    where_ $ user ^. UserAccountName ==. val name
+        &&. fromCrypt (user ^. UserAccountPasswordHash) pwd
+    pure user
+@
+
+-- @since 3.5.1
+-}
+fromCrypt :: SqlString s => SqlExpr (Value s) -> s -> SqlExpr (Value Bool)
+fromCrypt expr pass =
+    expr
+        ==. ERaw
```

Comment:
why is this including ==.

Could we instead just expose crypt and gen_salt as functions?

Reply:
Hi, this comes mostly from the fact that I can't replicate pg's `crypt()` overloaded behavior in haskell in order to only have  `crypt` and `genSalt` functions. A function 

````haskell
crypt :: SqlString s => s -> SqlExpr (Value s) -> SqlExpr (Value s)
````
would need a context switching parameter to identify when it is used for storing passwords and when for retrieving passwords.
 
With this in mind, `crypt()` always needs to use `gen_salt()` when used to store passwords, and so I think it makes sense to embed it in `toCrypt` instead of always including it explicitly. Moreover it protects the user from falsely using always the same salt by passing something like `val "salt"`.  

`fromCrypt` could skip the `==.` but then every time it was used, the user would have to duplicate the same persist field:

```haskell
login name pwd = select $ do
   user <- from $ Table UserAccount
   where_ $ user ^. UserAccountName ==. val name
   -- ^ as before
         &&. user ^. UserAccountPasswordHash ==.  fromCrypt pwd  (user ^. UserAccountPasswordHash)
````
Having the `==.` inside `fromCrypt` makes its use a bit more concise. 

Having said that, I understand that this might not be generic enough for esqueleto, so I would be happy to try to expose genSalt as a function and change `fromCrypt` if you think it would be better for the library.

btw, I've seen that I missed some parameters in some hashing algorithms. I'll add these as well.

Thanks for the help

Reply:
As a naming convention, I'd probably want `fromCrypt` to perform the decryption, and another function `checkEncrypted` that does the helper stuff. So your code example would be:

```haskell
    && user ^. UserAccountPasswordHash ==. fromCrypt pwd (user ^. UserAccountPasswordHash)
   -- or
    && encryptedEquals (user ^. UserAccountPasswordHash) pwd
```